docs/CppInteroperability/Roadmap.md

```diff
@@ -0,0 +1,141 @@
+
+```
+var v = vector(1) 
+let start = v.begin() 
+doSomething(start) 
+fixLifetime(v)
+```
+
+Here, because Swift inserts a copy before the call to begin, v projects a dangling reference. This is an example of how subtly different lifetime models make using C++ types from Swift hard, if their semantics aren’t understood by the compiler. 
+
+To make these APIs safe and usable, Swift cannot import unsafe projections of types that own memory, because they don’t fit the Swift model. Instead, the Swift compiler can try to infer what, semantically, the API is trying to do, or the library author can provide this information via annotations. In this case, the Swift compiler can infer that begin returns an iterator, which Swift can represent through the existing, safe Swift iterator interface. In the future, Swift may gain the necessary ownership features to import and use projections of types that own memory in a safe way, but today, this is not possible, and importing these projections would result in a *less safe* API in Swift than in C++ (violating the first goal above). Other than methods which project owned memory, owned types, their instances, methods on owned types, and other APIs that use owned types are generally considered to be safe and usable.
+
+Note: in the example above, “start” is a pointer type. The use of this pointer in the “begin” API is unsafe, but the type of start itself is not unsafe. In other words, safety restrictions need not be applied to pointer types themselves but rather their unsafe uses. 
+
+### Iterators
+
+Because iterators and ranges follow some specific API semantics, the Swift compiler can map them to a safe, ergonomic, native interface which Egor Zhdan will describe here :)
+
+### Generic APIs
+
+C++ provides a couple of tools for writing generic APIs: templates, concepts, virtual classes (inheritance), and various combinations and permutations of these. Templates are likely the most common tool for creating generic APIs. Unfortunately, as outlined in [this forum post](https://forums.swift.org/t/bridging-c-templates-with-interop/55003) (Bridging C++ Templates with Interop), C++ templates do not map cleanly to Swift generics, making interoperability between these generic APIs extremely difficult. Despite this, the linked forum post proposes various strategies for importing C++ templates derived from goals similar to the ones outlines in this roadmap. These proposals should probably be factored into this section of the roadmap at some point in the future.
```

Comment:
```suggestion
C++ provides a few of tools for writing generic APIs: templates, concepts, virtual classes (inheritance), and various combinations and permutations of these. Templates are likely the most common tool for creating generic APIs. Unfortunately, as outlined in [this forum post](https://forums.swift.org/t/bridging-c-templates-with-interop/55003) (Bridging C++ Templates with Interop), C++ templates do not map cleanly to Swift generics, making interoperability between these generic APIs extremely difficult. Despite this, the linked forum post proposes various strategies for importing C++ templates derived from goals similar to the ones outlines in this roadmap. These proposals should probably be factored into this section of the roadmap at some point in the future.
```

Reply:
Do we want to explicitly call out issues with type dependent and value dependent templates?  I suspect that the latter is something we will need to prohibit from being instantiated by Swift code, but may be able to work with `extern template` cases of.

Reply:
Yes. I would like to inline much of that forum post into this roadmap, but it will require some re-writing and I wanted to get the roadmap out ASAP. Anyway, yes, we should add that here.